### PR TITLE
Support IPEX backend

### PR DESCRIPTION
**Pull request by violetch24**
# What does this PR do?
This PR is used to support INC IPEX features and add QA example.

**examples/neural_compressor/question-answering/run_qa.py**
- Change `calib_dataloader`  from `get_train_dataloader()` to `get_eval_dataloader()` when using INC IPEX backend, in order to keep the form of `example_inputs` identical.

**optimum/intel/neural_compressor/optimization.py**
- Load `pytorch_model.bin` when using INC IPEX backend.

**optimum/intel/neural_compressor/quantization.py**
- Use `torch.jit.load` to recover the int8 model quantized by INC IPEX backend, instead of `torch.load`.

**examples/neural_compressor/question-answering/README.md**
- Add INC IPEX backend example explanation

**examples/neural_compressor/test_examples.py**
- Add a unit test for `run_qa.py` using INC IPEX backend

**Example**:
The following example applies INC IPEX backend:
python run_qa.py \
    --model_name_or_path bert-large-uncased-whole-word-masking-finetuned-squad \
    --dataset_name squad \
    --apply_quantization \
    --quantization_approach static \
    --do_train \
    --do_eval \
    --verify_loading \
    --output_dir /tmp/squad_output \
    --backend ipex

**Environment for this PR:**
Python 3.8.13
transformers 4.24.0
datasets 2.6.1
intel-extension-for-pytorch 1.12.300
torch 1.12.0

### CODE REVIEW

**Review by hshen14 — APPROVED** [review]
LGTM

**Comment by violetch24** [comment]
@echarlaix Hi Ella, please review. Thanks!

File: examples/neural_compressor/question-answering/README.md [hunk 1]
```diff
@@ -55,6 +55,8 @@ python run_qa.py \
     --output_dir /tmp/squad_output
 ```
 
+The flag `--backend` can be passed along and set to `ipex` to use INC IPEX backend. Default backend is `pytorch_fx` if value not specified.
```
Comment (echarlaix) [inline examples/neural_compressor/question-answering/README.md:58]: Why not use a flag `--use_ipex` to make it clear that it adds the capacity to use the ipex backend ? Also should add more information about the gains ?

File: examples/neural_compressor/question-answering/run_qa.py [hunk 2]
```diff
@@ -914,7 +921,13 @@ def train_func(model):
     if optim_args.apply_quantization and optim_args.verify_loading:
 
         # Load the model obtained after Intel Neural Compressor quantization
-        loaded_model = IncQuantizedModelForQuestionAnswering.from_pretrained(training_args.output_dir)
+        samples = next(iter(trainer.get_eval_dataloader()))
+        example_inputs = []
+        for key, value in samples.items():
+            example_inputs.append(samples[key])
+        loaded_model = IncQuantizedModelForQuestionAnswering.from_pretrained(
+            training_args.output_dir, example_inputs=example_inputs
```
Comment (echarlaix) [inline examples/neural_compressor/question-answering/run_qa.py:929]: We currently don't need an `example_inputs` to load the model, but maybe you wanted to add it to the `IncQuantizer` ?

File: optimum/intel/neural_compressor/quantization.py [hunk 3]
```diff
@@ -273,6 +273,10 @@ def from_pretrained(
         else:
             model_class._keys_to_ignore_on_load_missing.extend(missing_keys_to_ignore_on_load)
 
+        example_inputs = kwargs["example_inputs"] if "example_inputs" in kwargs.keys() else None
+        if "example_inputs" in kwargs.keys():
+            del kwargs["example_inputs"]
```
Comment (echarlaix) [inline optimum/intel/neural_compressor/quantization.py:278]:
```suggestion
        example_inputs = kwargs.pop("example_inputs", None)
```

File: optimum/intel/neural_compressor/quantization.py [hunk 4]
```diff
@@ -320,7 +324,23 @@ def from_pretrained(
 
                     raise EnvironmentError(msg)
 
-            state_dict = torch.load(state_dict_path)
+            try:
+                state_dict = torch.jit.load(state_dict_path)
+            except:
+                state_dict = torch.load(state_dict_path)
+        if isinstance(state_dict, torch.jit._script.RecursiveScriptModule):
+            assert example_inputs, "Please provide the example_inputs for IncQuantizedModel loading."
```
Comment (echarlaix) [inline optimum/intel/neural_compressor/quantization.py:332]: Can you raise an exception instead ? (to keep consistency)

File: optimum/intel/neural_compressor/quantization.py [hunk 5]
```diff
@@ -320,7 +324,23 @@ def from_pretrained(
...
+            q_model = torch.jit.freeze(stat_dict.eval())
+            # After freezing, run 1 time to warm up the profiling graph executor to insert prim::profile
+            # At the 2nd run, the llga pass will be triggered and the model is turned int
+            # an int8 model: prim::profile will be removed and will have LlgaFusionGroup in the graph
+            if isinstance(example_inputs, torch.Tensor):
```
Comment (echarlaix) [inline optimum/intel/neural_compressor/quantization.py:337]: It seems a little bit counterintuitive to have to give an example inputs to load the model, can we make it optional ? (and feed a dummy inputs by default ?)

File: optimum/intel/neural_compressor/quantization.py [hunk 6]
```diff
@@ -320,7 +324,23 @@ def from_pretrained(
 
                     raise EnvironmentError(msg)
 
-            state_dict = torch.load(state_dict_path)
+            try:
```
Comment (echarlaix) [inline optimum/intel/neural_compressor/quantization.py:327]: I think we can add a flag to specify whether or not to load a ScriptModule instead ?

File: examples/neural_compressor/question-answering/README.md (hunk 1 above)
Comment (violetch24) [reply examples/neural_compressor/question-answering/README.md:58]: We now use the flag `--use_ipex`. We set `config.use_ipex` to `False` in `train()` and `evaluate()` in `trainer.py` to avoid bf16 issues when calling `trainer.train()` and `trainer.evaluate()`.

File: optimum/intel/neural_compressor/quantization.py (hunk 6 above)
Comment (violetch24) [reply optimum/intel/neural_compressor/quantization.py:325]: We added a config to specify the way to load.

File: optimum/intel/neural_compressor/quantization.py (hunk 4 above)
Comment (violetch24) [reply optimum/intel/neural_compressor/quantization.py:330]: Done

File: optimum/intel/neural_compressor/quantization.py (hunk 5 above)
Comment (violetch24) [reply optimum/intel/neural_compressor/quantization.py:335]: We now feed dummy inputs instead of example inputs. Here two inferences are required in order to replace all int8 ops.

File: examples/neural_compressor/question-answering/run_qa.py (hunk 2 above)
Comment (violetch24) [reply examples/neural_compressor/question-answering/run_qa.py:929]: We don't need to add `example_inputs` to `IncQuantizer` because we can get it from calibration dataloader.
We now use `model.dummy_inputs` as example_inputs according to your suggestion, so we don't need it for loading function now.

File: examples/neural_compressor/question-answering/README.md [hunk 7]
```diff
@@ -55,6 +55,8 @@ python run_qa.py \
     --output_dir /tmp/squad_output
 ```
 
+The flag `--use_ipex` can be passed along to use INC IPEX backend. Default backend is `pytorch_fx` if value not specified.
+
```
Comment (PenghuiCheng) [inline examples/neural_compressor/question-answering/README.md:59]: Please point out that IPEX only supports static quantization now.

File: examples/neural_compressor/question-answering/README.md [hunk 8]
```diff
@@ -55,6 +55,8 @@ python run_qa.py \
     --output_dir /tmp/squad_output
 ```
 
+The flag `--use_ipex` can be passed along to use INC IPEX backend. Default backend is `pytorch_fx` if value not specified. INC IPEX only supports static quantization for now.
```
Comment (echarlaix) [inline examples/neural_compressor/question-answering/README.md:58]: Do you think we should add an extra comment on why it's interesting to use IPEX ? Also I think we should raise an error if the user tries a combination which is not supported (IPEX + dynamic quantization for example)

File: optimum/intel/neural_compressor/quantization.py [hunk 9]
```diff
@@ -279,6 +279,8 @@ def from_pretrained(
 
         model = model_class.from_pretrained(model_name_or_path, **kwargs)
 
+        dummy_inputs = list(model.dummy_inputs.keys())
+
```
Comment (PenghuiCheng) [inline optimum/intel/neural_compressor/quantization.py:283]: Is it dummy_inputs.values()?

Comment (violetch24) [reply optimum/intel/neural_compressor/quantization.py:285]: fixed

File: examples/neural_compressor/question-answering/README.md (hunk 8 above)
Comment (violetch24) [reply examples/neural_compressor/question-answering/README.md]: Explanation and warning added.

**Review by echarlaix — APPROVED** [review]
Thanks a lot for adding this @violetch24
